### PR TITLE
Use ColumnDecorator to configure LabKey "column types" 

### PR DESCRIPTION
**Pull request by labkey-matthewb**
#### Rationale
We inconsistently set up certain common "column types" e.g. ParticipantId or Container.  Use the existing ConceptURI property to register ColumnDecorators that configure these columns.

#### Related Pull Requests
https://github.com/LabKey/accounts/pull/142
https://github.com/LabKey/commonAssays/pull/316
https://github.com/LabKey/compliance/pull/107
https://github.com/LabKey/dataintegration/pull/98
https://github.com/LabKey/DiscvrLabKeyModules/pull/98
https://github.com/LabKey/ehrModules/pull/146
https://github.com/LabKey/MaxQuant/pull/37
https://github.com/LabKey/platform/pull/2121
https://github.com/LabKey/sampleManagement/pull/528

### CODE REVIEW

File: api/src/org/labkey/api/data/AbstractTableInfo.java [hunk 1]
```diff
@@ -655,6 +667,49 @@ public MutableColumnInfo addColumn(MutableColumnInfo column)
         return column;
     }
 
+    /**
+     * This method can be used to to replace the implementation of a column during construction.
+     * This is usually only done in TableInfo.afterConstruct() to modify the behavior of a column.
+     * Because the ColumnInfo implementation can change in afterConstruct(), TableInfo implementations
+     * should not hold to columnInfo references by FieldKey, and not by reference.
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/data/AbstractTableInfo.java]: Should this say:

```suggestion
     * should hold to columnInfo references by FieldKey, and not by reference.

```

File: api/src/org/labkey/api/query/column/BuiltInColumnTypes.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * These standard columns are not module specific, and need to be recognized by SchemaTableInfo * /
+ */
+public enum BuiltInColumnTypes
+{
+    Container(When.insert, JdbcType.GUID, BuiltInColumnTypes.CONTAINERID_CONCEPT_URI, null),
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/query/column/BuiltInColumnTypes.java:17]: This is great.  Should Owner and LSID be included?

File: api/src/org/labkey/api/query/column/ConceptURIColumnInfoTransformer.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import org.labkey.api.data.ColumnInfo;
+import org.labkey.api.data.MutableColumnInfo;
+import org.labkey.api.data.WrappedColumnInfo;
+
+public interface ConceptURIColumnInfoTransformer extends ColumnInfoTransformer, MutableColumnInfoTransformer
+{
+    @NotNull String getConceptURI();
+
+    // ColumnInfoTransformer provides only for wrapping a ColumnInfo
+    // In some cases it sense to allow mutating (AbstractTableInfo.afterConstruct())
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/query/column/ConceptURIColumnInfoTransformer.java]:
```suggestion
    // In some cases it makes sense to allow mutating (AbstractTableInfo.afterConstruct())
```

File: api/src/org/labkey/api/view/template/PageConfig.java [hunk 4]
```diff
@@ -92,7 +92,14 @@
     private boolean _includeLoginLink = true;
     private boolean _includeSearch = true;
     private int _minimumWidth = 400;
-    private LinkedHashSet<ClientDependency> _resources = new LinkedHashSet<>();
+    private LinkedHashSet<ClientDependency> _resources = new LinkedHashSet<>()
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/view/template/PageConfig.java]: Left over from debugging?

File: query/src/org/labkey/query/QueryModule.java [hunk 5]
```diff
@@ -137,6 +138,7 @@
     public QueryModule()
     {
         QueryService.setInstance(new QueryServiceImpl());
+        BuiltInColumnTypes.registerStandardColumnDecorators();
```
Comment (labkey-kevink) [inline query/src/org/labkey/query/QueryModule.java]:
```suggestion
        BuiltInColumnTypes.registerStandardColumnTransformers();

```

File: api/src/org/labkey/api/query/column/BuiltInColumnTypes.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        names = Collections.unmodifiableMap(map);
+    }
+
+    public static BuiltInColumnTypes findBuiltInType(ColumnInfo col)
+    {
+        BuiltInColumnTypes type = names.get(col.getName());
+        return null != type && type.matches(col) ? type : null;
+    }
+
+    public static void registerStandardColumnDecorators()
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/query/column/BuiltInColumnTypes.java]:
```suggestion
    public static void registerStandardColumnTransformers()
```

File: query/src/org/labkey/query/QueryServiceImpl.java [hunk 7]
```diff
@@ -3182,6 +3184,26 @@ public QueryAnalysisService getQueryAnalysisService()
         return _queryAnalysisService;
     }
 
+
+
+
+    /* registry of ColumnDecorators use to build common columns */
```
Comment (labkey-kevink) [inline query/src/org/labkey/query/QueryServiceImpl.java]:
```suggestion
    /* registry of ColumnInfoTransformer use to build common columns */
```

File: api/src/org/labkey/api/query/QueryService.java [hunk 8]
```diff
@@ -586,4 +589,26 @@ public JSONObject toJSON()
...
+    default void registerColumnInfoTransformer(@NotNull ConceptURIColumnInfoTransformer t)
+    {
+        registerColumnInfoTransformer(t.getConceptURI(), t);
+    }
+
+    void registerColumnInfoTransformer(@NotNull String uri, @NotNull ColumnInfoTransformer t);
+
+    ColumnInfoTransformer findColumnInfoTransformer(String conceptURI);
+
+    default MutableColumnInfo applyColumnDecorator(MutableColumnInfo col)
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/query/QueryService.java]:
```suggestion
    default MutableColumnInfo applyColumnTransformer(MutableColumnInfo col)
```

File: api/src/org/labkey/api/query/UserIdQueryForeignKey.java [hunk 9]
```diff
@@ -77,19 +77,24 @@ private TableInfo createLookupTableInfo()
...
-        ret.setLocked(true);
+        if (null != ret)
+            ret.setLocked(true);
         return ret;
     }
 
-    /* set foreign key and display column */
+    /**
+     * set foreign key and display column
+     * Deprecated: relying on UserIdColumnDecorator is preferred()
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/query/UserIdQueryForeignKey.java]:
```suggestion
     * Deprecated: relying on UserIdColumnInfoTransformer is preferred()

```

File: query/src/org/labkey/query/QueryServiceImpl.java [hunk 10]
```diff
@@ -3182,6 +3184,26 @@ public QueryAnalysisService getQueryAnalysisService()
...
+    Map<String, ColumnInfoTransformer> columnTransformerMap = Collections.synchronizedMap(new CaseInsensitiveHashMap<>());
+
+    @Override
+    public void registerColumnInfoTransformer(@NotNull String uri, @NotNull ColumnInfoTransformer t)
+    {
+        if (null != columnTransformerMap.put(uri, t))
+            throw new ConfigurationException("More than one ColumnDecorator registered for " + uri);
```
Comment (labkey-kevink) [inline query/src/org/labkey/query/QueryServiceImpl.java]:
```suggestion
            throw new ConfigurationException("More than one ColumnTransformer registered for " + uri);
```

File: api/src/org/labkey/api/data/ContainerForeignKey.java [hunk 11]
```diff
@@ -29,11 +29,20 @@
  */
 public class ContainerForeignKey extends QueryForeignKey
 {
+    /**
+     * Deprecated: relying on ContainerIdColumnDecorator is preferred()
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/data/ContainerForeignKey.java]:
```suggestion
     * Deprecated: relying on ContainerIdColumnInfoTransformer is preferred()

```

File: api/src/org/labkey/api/data/ContainerForeignKey.java [hunk 12]
```diff
@@ -29,11 +29,20 @@
...
+     */
+     @Deprecated
     static public <COL extends MutableColumnInfo> COL initColumn(COL column, UserSchema schema)
     {
         return initColumn(column, schema, null);
     }
 
+
+    /**
+     * Deprecated: relying on ContainerIdColumnDecorator is preferred()
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/data/ContainerForeignKey.java]:
```suggestion
     * Deprecated: relying on ContainerIdColumnInfoTransformer is preferred()

```

File: api/src/org/labkey/api/query/column/ColumnDecorator.java [hunk 13]
```diff
@@ -0,0 +1,8 @@
+package org.labkey.api.query.column;
+
+import org.labkey.api.data.MutableColumnInfo;
+
+public interface ColumnDecorator
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/query/column/ColumnDecorator.java]: This class is no longer needed after the renaming.

**Review by labkey-kevink — APPROVED** [review]
Just a few lingering references to decorator, but otherwise looks good.

File: api/src/org/labkey/api/query/column/BuiltInColumnTypes.java (hunk 2 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/query/column/BuiltInColumnTypes.java:17]: Maybe.  This is the list used by SimpleTranslator.  We should look for more opportunities to use both BuiltInColumnTypes and ConceptURIColumnInfoTransformer.

**Pull request merged**